node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,893 @@
+
+		let msg = BitfieldGossipMessage {
+			relay_parent: hash_a.clone(),
+			signed_availability: signed.clone(),
+		};
+
+		let pool = sp_core::testing::SpawnBlockingExecutor::new();
+		let (mut ctx, mut handle) =
+			subsystem_test::make_subsystem_context::<BitfieldDistributionMessage, _>(pool);
+
+		let mut tracker = prewarmed_tracker(
+			validator.clone(),
+			signing_context.clone(),
+			msg.clone(),
+			vec![peer_b.clone()],
+		);
+
+		executor::block_on(async move {
+			let _res = handle_network_msg(
+				&mut ctx,
+				&mut tracker,
+				NetworkBridgeEvent::PeerMessage(peer_b.clone(), msg.encode()),
+			)
+			.timeout(Duration::from_millis(10))
+			.await
+			.expect("10ms is more than enough for sending messages. qed")
+			.expect("There are no error values. qed");
+
+			// we should have a reputiation change due to invalid signature
+			// @todo assess if Eq+PartialEq are viable to simplify this kind of code
+			if let AllMessages::NetworkBridge(NetworkBridgeMessage::ReportPeer(peer, rep)) =
+				handle.recv().await
+			{
+				assert_eq!(peer, peer_b);
+				assert_eq!(rep, COST_SIGNATURE_INVALID);
+			} else {
+				panic!("Received unexpected message type.");
+			}
+		});
+	}
+
+	#[test]
+	fn receive_invalid_validator_index() {
+		let _ = env_logger::builder()
+			.filter(None, log::LevelFilter::Trace)
+			.is_test(true)
+			.try_init();
+
+		let hash_a: Hash = [0; 32].into();
+		let hash_b: Hash = [1; 32].into(); // other
+
+		let peer_a = PeerId::random();
+		let peer_b = PeerId::random();
+		assert_ne!(peer_a, peer_b);
+
+		let signing_context = SigningContext {
+			session_index: 1,
+			parent_hash: hash_a.clone(),
+		};
+
+		// validator 0 key pair
+		let (validator_pair, _seed) = ValidatorPair::generate();
+		let validator = validator_pair.public();
+
+		let payload = AvailabilityBitfield(bitvec![bitvec::order::Lsb0, u8; 1u8; 32]);
+		let signed =
+			Signed::<AvailabilityBitfield>::sign(payload, &signing_context, 42, &validator_pair);
+
+		let msg = BitfieldGossipMessage {
+			relay_parent: hash_a.clone(),
+			signed_availability: signed.clone(),
+		};
+
+		let pool = sp_core::testing::SpawnBlockingExecutor::new();
+		let (mut ctx, mut handle) =
+			subsystem_test::make_subsystem_context::<BitfieldDistributionMessage, _>(pool);
+
+		let mut tracker = prewarmed_tracker(
+			validator.clone(),
+			signing_context.clone(),
+			msg.clone(),
+			vec![peer_b.clone()],
+		);
+
+		executor::block_on(async move {
+			let _res = handle_network_msg(
+				&mut ctx,
+				&mut tracker,
+				NetworkBridgeEvent::PeerMessage(peer_b.clone(), msg.encode()),
+			)
+			.timeout(Duration::from_millis(10))
+			.await
+			.expect("10ms is more than enough for sending messages. qed")
+			.expect("There are no error values. qed");
+
+			// we should have a reputiation change due to invalid signature
+			// @todo assess if Eq+PartialEq are viable to simplify this kind of code
```

Comment:
we could use `assert_matches` crate
https://github.com/paritytech/substrate/blob/59ee76a0f342ae0dc1c6a6777d24d7811d9261ef/client/network/src/light_client_handler.rs#L1568-L1570

Reply:
Excellent suggestion!